File: scripts/shared/manage_instructions.py
"""Manage template-owned AGENTS/CLAUDE instruction blocks."""

from __future__ import annotations

import argparse
from pathlib import Path


LEGACY_BLOCKS = [
    ("<!-- BEGIN BEADS INTEGRATION -->", "<!-- END BEADS INTEGRATION -->"),
    ("<!-- BEGIN BEADS INTEGRATION v:", "<!-- END BEADS INTEGRATION -->"),
    ("<!-- br-agent-instructions-v", "<!-- end-br-agent-instructions -->"),
    ("<!-- bv-agent-instructions-v", "<!-- end-bv-agent-instructions -->"),
    ("<!-- BEGIN TEMPLATE BR WORKFLOW -->", "<!-- END TEMPLATE BR WORKFLOW -->"),
]

LEGACY_SNIPPETS = [
    "## Issue Tracking\n\nUses `"
    + chr(98)
    + chr(100)
    + "` (beads/Dolt). See `AGENTS.md` for repo rules and `BEADS_WORKFLOW.md` for the planner/executor workflow. Never use markdown TODOs or alternate task trackers.\n",
    "## Workflow Guide\n\nUse `BEADS_WORKFLOW.md` for the current planner, manual executor, and swarm executor flow.\nAll workflow skills are repo-local: Codex skills live under `.codex/skills/`, Claude skills under `.claude/skills/`.\nPreferred entry points are `plan-beads`, `validate-beads`, `start-epic-worktree`, `swarm-epic`, and `executor-once`.\nUse `plan-debate` before `beads-planner` when the user asks for extra scrutiny or the plan is risky.\nUse `executor-loop` or `executor-loop-epic` for sequential autonomy when swarm coordination is not needed.\nThe executor test skill lives at `.codex/skills/build-and-test/SKILL.md`; use it between implementation and final verification.\nUse `scripts/windows/start-epic-worktree.ps1` or `scripts/posix/start-epic-worktree.sh` to prepare epic worktrees.\nUse `scripts/windows/workflow-status.ps1` or `scripts/posix/workflow-status.sh` to inspect `.beads/workflow/` plus the shared control plane.\nUse `scripts/windows/agent-mail.ps1` or `scripts/posix/agent-mail.sh` for shared epic locks, reservations, and mailbox inspection.\n\nImportant:\n\n- keep this section outside the Beads-managed `AGENTS.md` block\n- do not edit inside `<!-- BEGIN BEADS INTEGRATION --> ... <!-- END BEADS INTEGRATION -->`\n",
    "## Workflow Guide\n\nUse `BEADS_WORKFLOW.md` for the current planner, manual executor, and swarm executor flow.\nAll workflow skills are repo-local: Codex skills live under `.codex/skills/`, Claude skills under `.claude/skills/`.\nPreferred entry points are `plan-beads`, `validate-beads`, `start-epic-worktree`, `swarm-epic`, and `executor-once`.\nUse `planner-research` only inside a planner session when `brainstorming` still leaves material factual uncertainty.\nUse `plan-debate` before `beads-planner` when the user asks for extra scrutiny or the plan is risky.\nUse `executor-loop` or `executor-loop-epic` for sequential autonomy when swarm coordination is not needed.\nThe executor test skill lives at `.codex/skills/build-and-test/SKILL.md`; use it between implementation and final verification.\nUse `scripts/windows/start-epic-worktree.ps1` or `scripts/posix/start-epic-worktree.sh` to prepare epic worktrees.\nUse `scripts/windows/workflow-status.ps1` or `scripts/posix/workflow-status.sh` to inspect `.beads/workflow/` plus the shared control plane.\nUse `scripts/windows/agent-mail.ps1` or `scripts/posix/agent-mail.sh` for shared epic locks, reservations, and mailbox inspection.\n\nImportant:\n\n- keep this section outside the Beads-managed `AGENTS.md` block\n- do not edit inside `<!-- BEGIN BEADS INTEGRATION --> ... <!-- END BEADS INTEGRATION -->`\n",
    "## Workflow Guide\n\nUse `BEADS_WORKFLOW.md` for the current planner, manual executor, and swarm executor flow.\nAll workflow skills are repo-local: Codex skills live under `.codex/skills/`, Claude skills under `.claude/skills/`.\nPreferred entry points are `plan-beads`, `validate-beads`, `start-epic-worktree`, `swarm-epic`, and `executor-once`.\nUse `executor-loop` or `executor-loop-epic` for sequential autonomy when swarm coordination is not needed.\nThe executor test skill lives at `.codex/skills/build-and-test/SKILL.md`; use it between implementation and final verification.\nUse `scripts/windows/start-epic-worktree.ps1` or `scripts/posix/start-epic-worktree.sh` to prepare epic worktrees.\nUse `scripts/windows/workflow-status.ps1` or `scripts/posix/workflow-status.sh` to inspect `.beads/workflow/` plus the shared control plane.\nUse `scripts/windows/agent-mail.ps1` or `scripts/posix/agent-mail.sh` for shared epic locks, reservations, and mailbox inspection.\n\nImportant:\n\n- keep this section outside the Beads-managed `AGENTS.md` block\n- do not edit inside `<!-- BEGIN BEADS INTEGRATION --> ... <!-- END BEADS INTEGRATION -->`\n",
    "## Workflow Guide\n\nUse `BEADS_WORKFLOW.md` for the current planner, manual executor, and swarm executor flow.\nAll workflow skills are repo-local: Codex skills live under `.codex/skills/`, Claude skills under `.claude/skills/`.\nPreferred entry points are `plan-beads`, `validate-beads`, `start-epic-worktree`, `swarm-epic`, and `executor-once`.\nUse `planner-research` only inside a planner session when `brainstorming` still leaves material factual uncertainty.\nUse `executor-loop` or `executor-loop-epic` for sequential autonomy when swarm coordination is not needed.\nThe executor test skill lives at `.codex/skills/build-and-test/SKILL.md`; use it between implementation and final verification.\nUse `scripts/windows/start-epic-worktree.ps1` or `scripts/posix/start-epic-worktree.sh` to prepare epic worktrees.\nUse `scripts/windows/workflow-status.ps1` or `scripts/posix/workflow-status.sh` to inspect `.beads/workflow/` plus the shared control plane.\nUse `scripts/windows/agent-mail.ps1` or `scripts/posix/agent-mail.sh` for shared epic locks, reservations, and mailbox inspection.\n\nImportant:\n\n- keep this section outside the Beads-managed `AGENTS.md` block\n- do not edit inside `<!-- BEGIN BEADS INTEGRATION --> ... <!-- END BEADS INTEGRATION -->`\n",
]


def strip_block(content: str, start_marker: str, end_marker: str) -> str:
    while True:
        start = content.find(start_marker)
        if start == -1:
            return content
        end = content.find(end_marker, start)
        if end == -1:
            return content
        end += len(end_marker)
        before = content[:start].rstrip("\r\n")
        after = content[end:].lstrip("\r\n")
        if before and after:
            content = before + "\n\n" + after
        else:
            content = before or after


def normalize_legacy(content: str) -> str:
    for start, end in LEGACY_BLOCKS:
        content = strip_block(content, start, end)
    for snippet in LEGACY_SNIPPETS:
        content = content.replace(snippet, "")
    return content.strip()


def parse_markers(snippet: str) -> tuple[str, str]:
    lines = [line for line in snippet.splitlines() if line.strip()]
    start = next((line for line in lines if line.startswith("<!-- BEGIN ")), None)
    end = next((line for line in lines if line.startswith("<!-- END ")), None)
    if not start or not end:
        raise ValueError("snippet must contain <!-- BEGIN ... --> and <!-- END ... --> markers")
    return start, end


def upsert(target: Path, snippet_path: Path) -> None:
    snippet = snippet_path.read_text(encoding="utf-8").strip()
    start_marker, end_marker = parse_markers(snippet)
    if target.exists():
        content = normalize_legacy(target.read_text(encoding="utf-8"))
        content = strip_block(content, start_marker, end_marker).strip()
        merged = snippet if not content else f"{content}\n\n{snippet}"
    else:
        merged = snippet
    target.write_text(merged.rstrip() + "\n", encoding="utf-8")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("target")
    parser.add_argument("snippet")
    args = parser.parse_args()
    upsert(Path(args.target), Path(args.snippet))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: lib/manageInstructions.js
'use strict';

// Port of scripts/shared/manage_instructions.py — manage template-owned
// AGENTS/CLAUDE instruction blocks. The legacy block/snippet definitions and the
// strip/upsert semantics are reproduced verbatim so that already-scaffolded
// downstreams have their stale blocks scrubbed on the next `update`.

const fs = require('fs');

const LEGACY_BLOCKS = [
  ['<!-- BEGIN BEADS INTEGRATION -->', '<!-- END BEADS INTEGRATION -->'],
  ['<!-- BEGIN BEADS INTEGRATION v:', '<!-- END BEADS INTEGRATION -->'],
  ['<!-- br-agent-instructions-v', '<!-- end-br-agent-instructions -->'],
  ['<!-- bv-agent-instructions-v', '<!-- end-bv-agent-instructions -->'],
  ['<!-- BEGIN TEMPLATE BR WORKFLOW -->', '<!-- END TEMPLATE BR WORKFLOW -->'],
];

// Build a "## Workflow Guide" legacy snippet. The four historical variants differ
// only by whether the planner-research and/or plan-debate lines are present.
function workflowGuide({ plannerResearch, planDebate }) {
  const lines = [
    '## Workflow Guide',
    '',
    'Use `BEADS_WORKFLOW.md` for the current planner, manual executor, and swarm executor flow.',
    'All workflow skills are repo-local: Codex skills live under `.codex/skills/`, Claude skills under `.claude/skills/`.',
    'Preferred entry points are `plan-beads`, `validate-beads`, `start-epic-worktree`, `swarm-epic`, and `executor-once`.',
  ];
  if (plannerResearch) {
    lines.push(
      'Use `planner-research` only inside a planner session when `brainstorming` still leaves material factual uncertainty.',
    );
  }
  if (planDebate) {
    lines.push(
      'Use `plan-debate` before `beads-planner` when the user asks for extra scrutiny or the plan is risky.',
    );
  }
  lines.push(
    'Use `executor-loop` or `executor-loop-epic` for sequential autonomy when swarm coordination is not needed.',
    'The executor test skill lives at `.codex/skills/build-and-test/SKILL.md`; use it between implementation and final verification.',
    'Use `scripts/windows/start-epic-worktree.ps1` or `scripts/posix/start-epic-worktree.sh` to prepare epic worktrees.',
    'Use `scripts/windows/workflow-status.ps1` or `scripts/posix/workflow-status.sh` to inspect `.beads/workflow/` plus the shared control plane.',
    'Use `scripts/windows/agent-mail.ps1` or `scripts/posix/agent-mail.sh` for shared epic locks, reservations, and mailbox inspection.',
    '',
    'Important:',
    '',
    '- keep this section outside the Beads-managed `AGENTS.md` block',
    '- do not edit inside `<!-- BEGIN BEADS INTEGRATION --> ... <!-- END BEADS INTEGRATION -->`',
  );
  return lines.join('\n') + '\n';
}

const LEGACY_SNIPPETS = [
  '## Issue Tracking\n\nUses `bd` (beads/Dolt). See `AGENTS.md` for repo rules and `BEADS_WORKFLOW.md` for the planner/executor workflow. Never use markdown TODOs or alternate task trackers.\n',
  workflowGuide({ plannerResearch: false, planDebate: true }),
  workflowGuide({ plannerResearch: true, planDebate: true }),
  workflowGuide({ plannerResearch: false, planDebate: false }),
  workflowGuide({ plannerResearch: true, planDebate: false }),
];

function stripBlock(content, startMarker, endMarker) {
  for (;;) {
    const start = content.indexOf(startMarker);
    if (start === -1) return content;
    const endIdx = content.indexOf(endMarker, start);
    if (endIdx === -1) return content;
    const end = endIdx + endMarker.length;
    const before = content.slice(0, start).replace(/[\r\n]+$/, '');
    const after = content.slice(end).replace(/^[\r\n]+/, '');
    if (before && after) {
      content = before + '\n\n' + after;
    } else {
      content = before || after;
    }
  }
}

function normalizeLegacy(content) {
  for (const [start, end] of LEGACY_BLOCKS) {
    content = stripBlock(content, start, end);
  }
  for (const snippet of LEGACY_SNIPPETS) {
    content = content.split(snippet).join('');
  }
  return content.trim();
}

function parseMarkers(snippet) {
  const lines = snippet.split(/\r?\n/).filter((line) => line.trim());
  const start = lines.find((line) => line.startsWith('<!-- BEGIN '));
  const end = lines.find((line) => line.startsWith('<!-- END '));
  if (!start || !end) {
    throw new Error('snippet must contain <!-- BEGIN ... --> and <!-- END ... --> markers');
  }
  return [start, end];
}

function upsert(targetPath, snippetPath) {
  const snippet = fs.readFileSync(snippetPath, 'utf8').trim();
  const [startMarker, endMarker] = parseMarkers(snippet);
  let merged;
  if (fs.existsSync(targetPath)) {
    let content = normalizeLegacy(fs.readFileSync(targetPath, 'utf8'));
    content = stripBlock(content, startMarker, endMarker).trim();
    merged = content ? `${content}\n\n${snippet}` : snippet;
  } else {
    merged = snippet;
  }
  fs.writeFileSync(targetPath, merged.replace(/\s+$/, '') + '\n', 'utf8');
}

module.exports = { upsert, stripBlock, normalizeLegacy, parseMarkers, LEGACY_SNIPPETS };
